torch/csrc/jit/codegen/onednn/kernel.cpp

```diff
@@ -0,0 +1,243 @@
+          "inputs with inputId ",
+          inputId,
+          " is missing");
+      auto* value = tensorIdToValue_[inputId];
+
+      TORCH_CHECK(
+          value->node()->kind() == prim::Constant &&
+              value->type()->cast<TensorType>(),
+          "inputs with inputId ",
+          inputId,
+          " should be a Constant tensor");
+      constantValues_.emplace_back(value);
+
+      auto const_tensor = toIValue(value)->toTensor();
+      constantInputs_.emplace_back(const_tensor);
+    }
+  }
+}
+
+std::map<size_t, int64_t> LlgaKernel::initializeTensorIdToOccurence() const {
+  std::map<size_t, int64_t> tensorIdToOccurence;
+  for (auto& lt : partition_.get_in_ports()) {
+    auto inputId = lt.get_id();
+    std::map<size_t, int64_t>::iterator it(tensorIdToOccurence.find(inputId));
+    if (it != tensorIdToOccurence.end()) {
+      it->second++;
+    } else {
+      tensorIdToOccurence[inputId] = 1;
+    }
+  }
+  return tensorIdToOccurence;
+}
+
+ArgSpecs LlgaKernel::initializeInputSpecs(const TensorArgs& inputs) {
+  ArgSpecs inputSpecs;
+  inputSpecs.reserve(nPartitionInputs_);
+  GRAPH_DEBUG("Initializing graph input logical tensors");
+  std::map<size_t, int64_t> tensorIdToOccurence =
+      initializeTensorIdToOccurence();
+  for (size_t i = 0; i < nGraphInputs_; i++) {
+    auto spec = ArgSpec(graph_->inputs()[i]).supplementTensorInfo(inputs[i]);
+    initializedInputIds_.insert(spec.tid());
+    int64_t occurence = tensorIdToOccurence[spec.tid()];
+    inputSpecs.insert(inputSpecs.end(), occurence, spec);
+    runArgsIdx_.insert(runArgsIdx_.end(), occurence, i);
+  }
+
+  GRAPH_DEBUG("Initializing constant input tensors");
+  initializeConstantInputs();
+
+  TORCH_CHECK(
+      inputSpecs.size() + constantValues_.size() == nPartitionInputs_,
+      "Partition inputs are missing");
+
+  GRAPH_DEBUG(
+      "Concatenating constant input logical tensors to graph input "
+      "logical tensors");
+  for (Value* constant_value : constantValues_) {
+    inputSpecs.emplace_back(ArgSpec(constant_value));
+  }
+  return inputSpecs;
+}
+
+ArgSpecs LlgaKernel::initializeOutputSpecs() const {
+  ArgSpecs outputSpecs;
+  outputSpecs.reserve(nOutputs_);
+  for (size_t i = 0; i < nOutputs_; i++) {
+    auto spec = ArgSpec(graph_->outputs()[i]);
+    if (useOpaqueLayout(i))
+      spec = spec.any();
+    outputSpecs.emplace_back(spec);
+  }
+  return outputSpecs;
+}
+
+std::tuple<RunArgs, RunArgs>
+LlgaKernel::prepareRunArgs(const TensorArgs &inputs,
+                           TensorArgs &outputs) const {
+  RunArgs runInputs, runOutputs;
+  auto numInputs = runArgsIdx_.size();
+  for (size_t i = 0; i < numInputs; i++) {
+    auto spec = inputSpecs_[i];
+    auto input = inputs[runArgsIdx_[i]];
+    runInputs.push_back(
+        {spec.logical_tensor(), Engine::getEngine(), input.data_ptr()});
+  }
+  auto numConstantInputs = constantInputs_.size();
+  for (size_t i = 0; i < numConstantInputs; i++) {
```

Comment:
The storage of constant tensors is being reused in each run. 

As for logical tensors, it's convenient to create logical tensors for each run, because otherwise, memory-management would become complex.